bgpd/bgp_nexthop.c

```diff
@@ -905,59 +1142,157 @@ static int show_ip_bgp_nexthop_table(struct vty *vty, const char *name,
 		struct bgp_nexthop_cache_head (*tree)[AFI_MAX];
 		struct bgp_nexthop_cache *bnc;
 		bool found = false;
+		json_object *json_afi = NULL;
 
 		if (!str2prefix(nhopip_str, &nhop)) {
-			vty_out(vty, "nexthop address is malformed\n");
+			if (!json)
+				vty_out(vty, "nexthop address is malformed\n");
 			return CMD_WARNING;
 		}
 		tree = import_table ? &bgp->import_check_table
 				    : &bgp->nexthop_cache_table;
+		if (json)
+			json_afi = json_object_new_object();
 		frr_each (bgp_nexthop_cache, &(*tree)[family2afi(nhop.family)],
 			  bnc) {
 			if (prefix_cmp(&bnc->prefix, &nhop))
 				continue;
-			bgp_show_nexthop(vty, bgp, bnc, true);
+			bgp_show_nexthop(vty, bgp, bnc, true, json_afi);
 			found = true;
 		}
-		if (!found)
+		if (json)
+			json_object_object_add(
+				json,
+				(family2afi(nhop.family) == AFI_IP) ? "ipv4"
+								    : "ipv6",
+				json_afi);
+		if (!found && !json)
 			vty_out(vty, "nexthop %s does not have entry\n",
 				nhopip_str);
 	} else
-		bgp_show_nexthops(vty, bgp, import_table);
+		bgp_show_nexthops(vty, bgp, import_table, json, afi, detail);
 
 	return CMD_SUCCESS;
 }
 
-static void bgp_show_all_instances_nexthops_vty(struct vty *vty)
+static void bgp_show_all_instances_nexthops_vty(struct vty *vty,
+						json_object *json, afi_t afi,
+						bool detail)
 {
 	struct listnode *node, *nnode;
 	struct bgp *bgp;
+	const char *inst_name;
+	json_object *json_instance = NULL;
 
 	for (ALL_LIST_ELEMENTS(bm->bgp, node, nnode, bgp)) {
-		vty_out(vty, "\nInstance %s:\n",
-			(bgp->inst_type == BGP_INSTANCE_TYPE_DEFAULT)
-				? VRF_DEFAULT_NAME
-				: bgp->name);
-		bgp_show_nexthops(vty, bgp, false);
+		inst_name = (bgp->inst_type == BGP_INSTANCE_TYPE_DEFAULT)
+				    ? VRF_DEFAULT_NAME
+				    : bgp->name;
+		if (json)
+			json_instance = json_object_new_object();
+		else
+			vty_out(vty, "\nInstance %s:\n", inst_name);
+
+		bgp_show_nexthops(vty, bgp, false, json_instance, afi, detail);
+
+		if (json)
+			json_object_object_add(json, inst_name, json_instance);
 	}
 }
 
+#include "bgpd/bgp_nexthop_clippy.c"
+
+DEFPY(show_ip_bgp_nexthop_ipv4,
+	  show_ip_bgp_nexthop_ipv4_cmd,
+      "show [ip] bgp [<view|vrf> VIEWVRFNAME$vrf] nexthop ipv4 [A.B.C.D] [detail$detail] [json$uj]",
+      SHOW_STR
+      IP_STR
+      BGP_STR
+      BGP_INSTANCE_HELP_STR
+      "BGP nexthop table\n"
+      "BGP nexthop IPv4 table\n"
+      "IPv4 nexthop address\n"
+      "Show detailed information\n"
+      JSON_STR)
+{
+	int rc = 0;
+	json_object *json = NULL;
+	char *nhop_str = NULL;
+	int nh_idx = 0;
+
+	if (argv_find(argv, argc, "A.B.C.D", &nh_idx))
+		nhop_str = argv[nh_idx]->arg;
```

Comment:
With DEFPY, we don't need argv_find( ) actually.

Reply:
Fixed it